modules/discretedata/dataset.h

```diff
@@ -0,0 +1,209 @@
+*
+* Inviwo - Interactive Visualization Workshop
+*
+* Copyright (c) 2012-2018 Inviwo Foundation
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+* list of conditions and the following disclaimer.
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+* this list of conditions and the following disclaimer in the documentation
+* and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+* ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+*********************************************************************************/
+
+#pragma once
+
+#include <discretedata/discretedatamoduledefine.h>
+#include <inviwo/core/common/inviwo.h>
+#include <inviwo/core/datastructures/datatraits.h>
+
+#include "channels/bufferchannel.h"
+#include "connectivity/connectivity.h"
+
+namespace inviwo {
+namespace dd {
+
+typedef std::shared_ptr<Channel> SharedChannel;
+typedef std::shared_ptr<const Channel> SharedConstChannel;
+
+struct ChannelCompare {
+    bool operator()(const std::pair<std::string, GridPrimitive>& u,
+                    const std::pair<std::string, GridPrimitive>& v) const {
+        return (u.second < v.second) || u.first.compare(v.first) < 0;
+    }
+};
+// Map used for storing and querying channels by name and GridPrimitive type.
+typedef std::map<
+    std::pair<std::string, GridPrimitive>  // Hashing by both name and GridPrimitive type
+    ,
+    SharedConstChannel  // Shared channels, type information only as meta property
+    ,
+    ChannelCompare>  // Lesser operator on string-Primitve pairs
+    BaseChannelMap;
+
+class IVW_MODULE_DISCRETEDATA_API DataChannelMap {
+public:
+    DataChannelMap() = default;
+    virtual ~DataChannelMap() = default;
+
+    /** Default copy shares Channels */
+    DataChannelMap(const DataChannelMap& copy) = default;
+
+    // Methods
+public:
+    /** Add a new channel to the set
+    *   @param channel Pointer to data, takes memory ownership
+    *   @return Shared pointer for further handling
+    */
+    SharedChannel addChannel(Channel* channel);
```

Comment:
In an new think like this, I don't think we need to supply the raw pointer interface. the user should always use std::make_shared to create the channels anyway. So I think this overload can be removed.